### PR TITLE
validate_presence_of works with delegation

### LINKED ISSUE #765: validate_presence_of does not work in an internal object association in 2.8.0
Hello,

I have been trying to use the `validate_presence_of` to check for the presence of an association in an internal object (to whom the wrapping object delegates the get and set messages), but I keep getting a `NoMethodError: undefined method 'each' for nil:NilClass` error.

Here is what I have:

```ruby
class Project < ActiveRecord::Base
  has_and_belongs_to_many :students

  # The rest of the class is omitted for the sake of brevity.
end
```

```ruby
class ProjectBuilder::SelectStudentsStep
  include ActiveModel::Model

  attr_accessor :project
  delegate :students, :students=, to: :project

  validates :students, presence: true
end
```

```ruby
require 'rails_helper'

RSpec.describe ProjectBuilder::SelectStudentsStep, type: :model do
  describe 'validations' do
    subject { ProjectBuilder::SelectStudentsStep.new(project: create(:project)) }
    it { should validate_presence_of(:students) }
  end
end
```

And here is the stack trace:

```
NoMethodError:
       undefined method `each' for nil:NilClass
     # /Users/alexbrahastoll/.rvm/gems/ruby-2.2.2@lumiar/gems/activerecord-4.2.2/lib/active_record/associations/collection_association.rb:365:in `replace'
     # /Users/alexbrahastoll/.rvm/gems/ruby-2.2.2@lumiar/gems/activerecord-4.2.2/lib/active_record/associations/collection_association.rb:47:in `writer'
     # /Users/alexbrahastoll/.rvm/gems/ruby-2.2.2@lumiar/gems/activerecord-4.2.2/lib/active_record/associations/builder/association.rb:123:in `students='
     # ./app/models/project_builder/select_students_step.rb:5:in `students='
     # /Users/alexbrahastoll/.rvm/gems/ruby-2.2.2@lumiar/gems/shoulda-matchers-2.8.0/lib/shoulda/matchers/active_model/allow_value_matcher.rb:259:in `set_attribute_ignoring_range_errors'
     # /Users/alexbrahastoll/.rvm/gems/ruby-2.2.2@lumiar/gems/shoulda-matchers-2.8.0/lib/shoulda/matchers/active_model/allow_value_matcher.rb:254:in `set_attribute'
     # /Users/alexbrahastoll/.rvm/gems/ruby-2.2.2@lumiar/gems/shoulda-matchers-2.8.0/lib/shoulda/matchers/active_model/allow_value_matcher.rb:223:in `block in matches?'
     # /Users/alexbrahastoll/.rvm/gems/ruby-2.2.2@lumiar/gems/shoulda-matchers-2.8.0/lib/shoulda/matchers/active_model/allow_value_matcher.rb:220:in `each'
     # /Users/alexbrahastoll/.rvm/gems/ruby-2.2.2@lumiar/gems/shoulda-matchers-2.8.0/lib/shoulda/matchers/active_model/allow_value_matcher.rb:220:in `none?'
     # /Users/alexbrahastoll/.rvm/gems/ruby-2.2.2@lumiar/gems/shoulda-matchers-2.8.0/lib/shoulda/matchers/active_model/allow_value_matcher.rb:220:in `matches?'
     # /Users/alexbrahastoll/.rvm/gems/ruby-2.2.2@lumiar/gems/shoulda-matchers-2.8.0/lib/shoulda/matchers/active_model/disallow_value_matcher.rb:16:in `matches?'
     # /Users/alexbrahastoll/.rvm/gems/ruby-2.2.2@lumiar/gems/shoulda-matchers-2.8.0/lib/shoulda/matchers/active_model/validation_matcher.rb:56:in `disallows_value_of'
     # /Users/alexbrahastoll/.rvm/gems/ruby-2.2.2@lumiar/gems/shoulda-matchers-2.8.0/lib/shoulda/matchers/active_model/validate_presence_of_matcher.rb:126:in `matches?'
     # ./spec/models/project_builder/select_students_step_spec.rb:6:in `block (3 levels) in <top (required)>'
```

I can get and set students in a Project object via a ProjectBuilder::SelectStudentsStep object just fine in a rails console (and the rails presence validation is also working correctly).

I am using ruby 2.2.2, rails 4.2.2 and shoulda-matchers 2.8.0

Is this a supported use case? Am I doing something wrong here?

Thanks in advance!

### PR DESCRIPTION
**Pull request by maurogeorge**
If we try define a validation to a attribute defined on a delegated object we
will receive a `NoMethodError` if the delegated attribute expect a collection as
value and not `nil` to a empty value.

Change the definition of what is a collection, before we are checking on
`:has_many` and `:has_and_belongs_to_many` association, now we check only if the
default value of the object is a empty array, since this is the default behavior
of a empty collection on `ActiveRecord`.

closes #765

### CODE REVIEW

File: spec/unit/shoulda/matchers/active_model/validate_presence_of_matcher_spec.rb [hunk 1]
```diff
@@ -52,6 +52,42 @@
     end
   end
 
+  context "a ActiveModel that's validate a delegated attribute to a ActiveModel" do
```
Comment (mcmire) [inline spec/unit/shoulda/matchers/active_model/validate_presence_of_matcher_spec.rb]: What about:

"an ActiveModel with a validation on an attribute that delegates to an association in another ActiveModel"

File: spec/unit/shoulda/matchers/active_model/validate_presence_of_matcher_spec.rb [hunk 2]
```diff
@@ -52,6 +52,42 @@
...
+        validates_presence_of :children
+      end.new(parent: parent)
+    end
+
+    it 'accepts' do
+      expect(@model).to validate_presence_of(:children)
+    end
+  end
+
+  context "a ActiveModel that's validate a delegated attribute to a has_and_belongs_to_many association" do
```
Comment (mcmire) [inline spec/unit/shoulda/matchers/active_model/validate_presence_of_matcher_spec.rb]: What about:

"an ActiveModel with a validation on an attribute that delegates to a habtm association in another ActiveModel"

File: spec/unit/shoulda/matchers/active_model/validate_presence_of_matcher_spec.rb [hunk 3]
```diff
@@ -52,6 +52,42 @@
     end
   end
 
+  context "a ActiveModel that's validate a delegated attribute to a ActiveModel" do
+    before do
+      parent = define_active_model_class('Parent', accessors: [:children]).new
+      @model = define_active_model_class('Example', accessors: [:parent]) do
+        delegate :children, :children=, to: :parent
+        validates_presence_of :children
+      end.new(parent: parent)
```
Comment (mcmire) [inline spec/unit/shoulda/matchers/active_model/validate_presence_of_matcher_spec.rb]: Can you put the `.new(parent: parent)` on its own line? This looks a little funny.

File: spec/unit/shoulda/matchers/active_model/validate_presence_of_matcher_spec.rb [hunk 4]
```diff
@@ -52,6 +52,42 @@
...
+    end
+  end
+
+  context "a ActiveModel that's validate a delegated attribute to a has_and_belongs_to_many association" do
+    before do
+      define_model :child
+      parent = define_model :parent do
+        has_and_belongs_to_many :children
+        validates_presence_of :children
+      end.new
```
Comment (mcmire) [inline spec/unit/shoulda/matchers/active_model/validate_presence_of_matcher_spec.rb]: Can you put `.new` here on its own line?

Also there should be line breaks around methods that take blocks.

File: spec/unit/shoulda/matchers/active_model/validate_presence_of_matcher_spec.rb [hunk 5]
```diff
@@ -52,6 +52,42 @@
...
+      create_table 'children_parents', id: false do |t|
+        t.integer :child_id
+        t.integer :parent_id
+      end
+      @model = define_active_model_class('Example', accessors: [:parent]) do
+        delegate :children, :children=, to: :parent
+        validates_presence_of :children
+      end.new(parent: parent)
```
Comment (mcmire) [inline spec/unit/shoulda/matchers/active_model/validate_presence_of_matcher_spec.rb]: This `.new` should go on its own line, too.

**Comment by maurogeorge** [comment]
@mcmire thanks for the review :heart_decoration:

I updated the code following your comments, and now the :dog: is barking should I follow the @houndci recomendation?

I updated other sections of the same file too :smile: I keep this on a separated commit since is a different change.

**Comment by mcmire** [comment]
Oh I'm sorry by "place `.new` on its own line" I meant let's make another variable to hold the instance of the model on the next line.

**Comment by maurogeorge** [comment]
@mcmire no problem :smile: 
I updated the code.

**Comment by eparreno** [comment]
What happened with that fix? I having that issue with delegated methods.

**Comment by mcmire** [comment]
Hey folks. In an effort to lighten our load as maintainers and be able to serve you better in the future, the shoulda-matchers team is working on cleaning out the cobwebs in this repo by pruning the backlog. As there are few of us, there are a lot of items that will simply never earn our attention in a reasonable time frame, and rather than giving you an empty promise, we think it makes more sense to focus on more recent issues. That means, unfortunately, that we must close this PR.

Don't take this the wrong way: our aim is not to diminish the effort people have made or dismiss problems that have been raised. If you feel that we should reopen this PR, then please let us know so that we can reprioritize it. Thanks!

**Pull request closed without merging**